File: Assembly.py
instr = {"NOP":"00",
         "JMP":"01",
         "JIC":"02",
         "JIZ":"03",
         "LDA":"04",
         "STA":"05",
         "LDB":"06",
         "STB":"07",
         "PLA":"08",
         "PSA":"09",
         "PLB":"0A",
         "PSB":"0B",
         "PLO":"0C",
         "PLR":"0D",
         "PSR":"0E",
         "SSP":"0F",
         "LIA":"10",
         "LIB":"11",
         "JMS":"12",
         "JCS":"13",
         "JZS":"14",
         "JMR":"15",
         "JCR":"16",
         "JZR":"17",
         "OPR":"2",
         "OPA":"3",
         "OPB":"4",
         "OPO":"5",
         "OPS":"6",
         "HDR +":"FFFD",
         "HDR P":"FFFE",
         "HLT":"FFFF"}

operation = {"BUFF":"0",
             "NOT" :"1",
             "ROR" :"2",
             "ROL" :"3",
             "OR"  :"4",
             "NOR" :"5",
             "AND" :"6",
             "NAND":"7",
             "ADD" :"8",
             "SUB" :"9"}

def decodeNumber(number,variable):
    if (number[0] == "#"):
        return f"{number[1:]:>02}"
    elif (number[0] == "%"):
        return f"{hex(int(number[1:],2))[2:]:>02}".upper()
    elif (number[0] == "$"):
        return f"{hex(int(number[1:]))[2:]:>02}".upper()
    else:
        return f"{variable[number]:>02}".upper()

def decodeOneInstr(line,variable):
    out = ""
    curr = line.split(" ")
    if (curr[0] in ["OPR","OPA","OPB","OPO","OPS"]):
        if (curr[0] == "OPR"):
            par = decodeNumber(curr[2],variable)
            out += instr[curr[0]] + operation[curr[1]] + par
        else:
            out += instr[curr[0]] + operation[curr[1]] + "00"
    elif (curr[0] in ["HLT","SHO"]):
        out += instr[curr[0]]
    elif (curr[0] in ["PLA","PSA","PLB","PSB","NOP","PLO","JMS","JCS","JZS"]):
        out += instr[curr[0]] + "00"
    else:
        par = decodeNumber(curr[1],variable)
        out += instr[curr[0]] + par
    return out

def decodeVariable(line, variable):
    name = ""
    val = ""
    ind = 0
    step = 0
    for index,value in enumerate(line.replace(" ","")):
        if (value == "="):
            step += 1
        elif (value != "=" and step == 0):
            name += value
        elif (value != "=" and step == 1):
            val += value
            
    return name, decodeNumber(val, variable)

def assemble(file):
    with open(file,mode="r") as of:
        A = of.read()
    lines = A.split("\n")
    out = ""
    count = 0
    variable = {}
    for i in lines:
        curr = i.replace("\t","")
        if (curr == ""):
            continue
        if (curr[-1] == ":"):
            variable[i[:-1]] = hex(count)[2:]
        elif ("=" in curr):
            pass
        else:
            count += 1
    for i in lines:
        curr = i.replace("\t","")
        if (curr == ""):
            continue
        if ("=" in curr):
            A, B = decodeVariable(curr,variable)
            variable[A] = B
        elif not (curr[-1] == ":"):
            out += decodeOneInstr(curr,variable) + " "
    print(out)
    return out

if __name__ == "__main__":
    with open(r"Compiled\B&F.f8p","w") as F:
        F.write(assemble(r"Program\BackAndForth.txt"))

File: C#/Assembly.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace Fox8Assembly
{
	static internal class Assembly
	{
	    static Dictionary<string, string> variable = new Dictionary<string,string>();

		static Dictionary<string, string> INSTR = new Dictionary<string, string> {
			{ "NOP", "00" },
			{ "JMP", "01" },
			{ "JIC", "02" },
			{ "JIZ", "03" },
			{ "LDA", "04" },
			{ "STA", "05" },
			{ "LDB", "06" },
			{ "STB", "07" },
			{ "PLA", "08" },
			{ "PSA", "09" },
			{ "PLB", "0A" },
			{ "PSB", "0B" },
			{ "PLO", "0C" },
			{ "PLR", "0D" },
			{ "PSR", "0E" },
			{ "SSP", "0F" },
			{ "LIA", "10" },
			{ "LIB", "11" },
			{ "JMS", "12" },
			{ "JCS", "13" },
			{ "JZS", "14" },
			{ "JMR", "15" },
			{ "JCR", "16" },
			{ "JZR", "17" },
			{ "OPR", "2" },
			{ "OPA", "3" },
			{ "OPB", "4" },
			{ "OPO", "5" },
			{ "OPS", "6" },
			{ "HDR+", "FFFD" },
			{ "HDRP", "FFFE" },
			{ "HLT", "FFFF" },
		};

		static Dictionary<string, string> OP = new Dictionary<string, string> {
			{ "BUFF", "0" },
			{ "NOT",  "1" },
			{ "ROR",  "2" },
			{ "ROL",  "3" },
			{ "OR",   "4" },
			{ "NOR",  "5" },
			{ "AND",  "6" },
			{ "NAND", "7" },
			{ "ADD",  "8" },
			{ "SUB",  "9" }
		};

		public static string DecodeNumber(string number) {
			if (number[0] == '#') {
				return number.Substring(1);
			}
			else if (number[0] == '%') {
				return String.Format("{0:X2}", Convert.ToUInt64(number.Substring(1), 2));
			}
			else if (number[0] == '$') {
				return String.Format("{0:X2}", Convert.ToUInt64(number.Substring(1)));
			}
			else {
				return variable[number];
			}
		}

		public static string DecodeOneInstr(string line) {
			string[] lines = line.Split(' ');
			string output = "";
			if (lines[0] == "OPR") { output = INSTR[lines[0]] + OP[lines[1]] + DecodeNumber(lines[2]); }
			else if (lines[0] == "OPA" || lines[0] == "OPB" || lines[0] == "OPO" || lines[0] == "OPS") { output = INSTR[lines[0]] + OP[lines[1]] + "00"; }
			else if (lines[0] == "HLT" || lines[0] == "HDR+" || lines[0] == "HDRP") { output = INSTR[lines[0]]; }
			else if (lines[0] == "NOP" || lines[0] == "PLA" || lines[0] == "PSA" || lines[0] == "PLB" || lines[0] == "PSB" || lines[0] == "PLO" || lines[0] == "PLR" || lines[0] == "PSR"
			|| lines[0] == "JMS" || lines[0] == "JCS" || lines[0] == "JZS") { output = INSTR[lines[0]] + "00"; }
			else { output = INSTR[lines[0]] + DecodeNumber(lines[1]); }
			return output;
		}

		public static string DecodeFile(string filename) {
			string[] content = File.ReadAllText(filename).Replace("\t","").Replace("\r","").Split("\n");
			string output = "";
			int count = 0;
			foreach (string line in content) {
				if (line.Trim() == "") { continue; }
				if (line.EndsWith(":")) { variable.Add(line.Remove(line.Length - 1), String.Format("{0:X2}", count)); }
				else if (line.Contains("=") || line.StartsWith("/")) { continue; }
				else { count++; }
			}
			foreach (string line in content) {
				if (line.Trim() == "") { continue; }
				if (line.Contains("=")) {
					string[] info = line.Split('=');
					variable.Add(info[0].Trim(), DecodeNumber(info[1].Trim()));
				}
				else if (line.StartsWith("/") || line.EndsWith(":")) { continue; }
				else {
					output += DecodeOneInstr(line) + " ";
				}
			}
			return output;
		}

		public static ushort[] GenerateCode(string filename) {
			string[] prog = DecodeFile(filename).Split(" ");
			ushort[] code = new ushort[prog.Length - 1];
			for (int i = 0; i < prog.Length - 1; i++)
			{
				code[i] = Convert.ToUInt16(prog[i], 16);
			}
			return code;
		}
	}

	/*internal class Program
	{
		static void Main(string[] args)
		{
			Console.WriteLine(Assembly.DecodeFile(@"C:\Users\maxen\Documents\Computer\Emulated\Program\Multiplication.txt"));
		}
	}*/
}
